main/src/ca/uwaterloo/flix/language/phase/jvm/GenExpression.scala

```diff
@@ -38,20 +38,38 @@ object GenExpression {
 
   type Ref[T] = Array[T]
 
-  case class MethodContext(clazz: JvmType.Reference,
+  sealed trait MethodContext {
+
+    def clazz: JvmType.Reference
+
+    def entryPoint: Label
+
+    def lenv: Map[Symbol.LabelSym, Label]
+
+    def localOffset: Int
+
+  }
+
+  case class InstanceContext(clazz: JvmType.Reference,
+                             entryPoint: Label,
+                             lenv: Map[Symbol.LabelSym, Label],
+                             newFrame: InstructionSet, // [...] -> [..., frame]
+                             setPc: InstructionSet, // [..., frame, pc] -> [...]
+                             localOffset: Int,
+                             pcLabels: Vector[Label],
+                             pcCounter: Ref[Int]
+                            ) extends MethodContext
+
+  case class StaticContext(clazz: JvmType.Reference,
```

Comment:
After thinking, can you rename this to DirectContext and the other to EffectContext?